src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -262,7 +290,439 @@ private RouteFW resolveTarget(
+                                     .streamId(streamId)
+                                     .traceId(traceId)
+                                     .authorization(authorization)
+                                     .build();
+
+        receiver.accept(reset.typeId(), reset.buffer(), reset.offset(), reset.sizeof());
+    }
+
+    private void doSignal(
+        MessageConsumer receiver,
+        long routeId,
+        long streamId,
+        long traceId)
+    {
+        final SignalFW signal = signalRW.wrap(writeBuffer, 0, writeBuffer.capacity()).routeId(routeId)
+                                        .streamId(streamId)
+                                        .traceId(traceId)
+                                        .build();
+
+        receiver.accept(signal.typeId(), signal.buffer(), signal.offset(), signal.sizeof());
+    }
+
+    private int decodePacketType(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        final MqttPacketFixedHeaderFW packet = mqttPacketFixedHeaderRO.tryWrap(buffer, offset, limit);
+
+        if (packet != null)
+        {
+            final int length = packet.remainingLength();
+            final MqttPacketType packetType = MqttPacketType.valueOf(packet.typeAndFlags() >> 4);
+            final MqttServerDecoder decoder = decodersByPacketType.getOrDefault(packetType, decodeError);
+
+            if (limit - packet.limit() >= length)
+            {
+                server.decoder = decoder;
+            }
+        }
+
+        return offset;
+    }
+
+    private int decodeConnect(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final long budgetId,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        MqttConnectFW mqttConnect = mqttConnectRO.tryWrap(buffer, offset, limit);
+
+        int progress = offset;
+        int reasonCode = 0x00;
+        if (mqttConnect == null || (mqttConnect.flags() & 0b0000_0001) != 0b0000_0000)
+        {
+            reasonCode = 0x81; // malformed packet
+        }
+        else if (!"MQTT".equals(mqttConnect.protocolName().asString()) || mqttConnect.protocolVersion() != 5)
+        {
+            reasonCode = 0x84; // unsupported protocol version
+        }
+        else if (server.connected)
+        {
+            reasonCode = 0x82; // protocol error
+        }
+
+        if (reasonCode == 0)
+        {
+            server.onDecodeConnect(traceId, authorization, reasonCode, mqttConnect);
+            server.decoder = decodePacketType;
+            progress = mqttConnect.limit();
+        }
+        else
+        {
+            server.doEncodeConnack(traceId, authorization, reasonCode);
+            server.doNetworkEnd(traceId, authorization);
```

Comment:
Use `server.onDecodeError` here too, then use `connected` to determine whether `onDecodeError` should `doEncodeConnack` or `doEncodeDisconnect`.